src/tests/test_server_stream.cpp

```diff
@@ -0,0 +1,520 @@
+            auto self = shared_from_this();
+            /* no yield! */ net::async_read(m_client.stream(), m_client.buffer(),
+                                            [this, self](const error_code &ec, size_t)
+               {
+               m_result.stop_timer();
+               // check that "waiting" for the server's shutdown turns out to be aborted
+               m_result.confirm("async_read is aborted", ec == net::error::operation_aborted);
+               });
+
+            m_result.confirm("received close_notify", !m_client.stream().shutdown_received());
+            m_client.stream().lowest_layer().close();
+            }
+         }
+
+      Result result() { return m_result.result(); }
+
+   private:
+      Client m_client;
+      Result_Wrapper m_result;
+      std::shared_ptr<Server> m_server;
+   };
+
+/* In this test case the server shuts down the connection but the client doesn't send the corresponding close_notify
+ * response. Instead, it closes the socket immediately.
+ * The server should see a short-read error.
+ */
+class Test_No_Shutdown_Response : public net::coroutine, public std::enable_shared_from_this<Test_No_Shutdown_Response>
+   {
+   public:
+      Test_No_Shutdown_Response(net::io_context& ioc, std::shared_ptr<Server> server)
+         : m_client(ioc),
+           m_result(ioc, "Test No Shutdown Response"),
+           m_server(server) {}
+
+      void run(const error_code& ec)
+         {
+         static auto test_case = &Test_No_Shutdown_Response::run;
+         reenter(*this)
+            {
+            m_result.set_timer("connect");
+            yield net::async_connect(m_client.stream().lowest_layer(), k_endpoints,
+                                     std::bind(test_case, shared_from_this(), _1));
+            m_result.check_ec("connect", ec);
+
+            m_result.set_timer("handshake");
+            yield m_client.stream().async_handshake(Botan::TLS::Connection_Side::CLIENT,
+                                                    std::bind(test_case, shared_from_this(), _1));
+            m_result.check_ec("handshake", ec);
+
+            m_server->shutdown();
+
+            m_result.set_timer("read close_notify");
+            yield net::async_read(m_client.stream(), m_client.buffer(),
+                                  std::bind(test_case, shared_from_this(), _1));
+            m_result.confirm("read gives EOF", ec == net::error::eof);
+            m_result.confirm("received close_notify", m_client.stream().shutdown_received());
+
+            m_result.stop_timer();
+
+            // close the socket rather than shutting down
+            m_server->expect_short_read();
+            m_client.stream().lowest_layer().close();
+            }
+         }
+
+      Result result() { return m_result.result(); }
+
+   private:
+      Client m_client;
+      Result_Wrapper m_result;
+      std::shared_ptr<Server> m_server;
+   };
+
+#include <boost/asio/unyield.hpp>
+
+template<typename TestT>
```

Comment:
If you find a more elegant solution than my test case classes, I'll be very grateful.